draft-ietf-quic-recovery.md

```diff
@@ -398,8 +417,80 @@ A packet sent on an RTO alarm MUST NOT be blocked by the sender's congestion
 controller. A sender MUST however count these bytes as additional bytes in
 flight, since this packet adds network load without establishing packet loss.
 
+## Multiple Packet Number Space Optimizations {#optimizations}
+
+There are cases where one may be able to gain recovery information from
+acknowledgements of packets in another packet number space, but they rely
+on complex assumptions about the peer’s processing and acknowledgement
+algorithms.  Even those are unable to quickly recover from cases such as
+losing the client's Initial, but receiving the 0-RTT packets.  Below are
+three different optimizations in increasing complexity that minimize
+handshake latency.
+
+### Empty ACK frames
+
+The EMPTY_ACK frame enables faster recovery of lost Initial and Handshake
+packets when the only other outstanding packets are undecryptable.
+In that case, the peer can’t decrypt the packet or packet number,
+but they do know a packet for a QUIC connection of a supported version was
+received.
+
+The receiver SHOULD send an EMPTY_ACK frame soon (e.g., 1ms) after
+undecryptable packets are received, even if those received packets are
+not buffered for later decryption.  The small delay allows for cases when
+0-RTT packets are reordered in front of the Initial, which is not uncommon
+on networks that prioritize small packets.  The receiver should limit the
+number of EMPTY_ACK frames sent to one per packet number space per RTT.
+
+When an EMPTY_ACK frame is received, a sender should immediately
+retransmit the missing handshake packet(s) as though the handshake timer
+fired and re-arm the handshake timer when the handshake packets are sent.
+If the missing handshake data was timer retransmitted after the packets
+that triggered the empty ack, then the empty ack should be ignored.
+
+An EMPTY_ACK frame does not acknowledge a new packet, and in cases
+when multiple packets are outstanding, the RTT signal is ambiguous,
+so it should not be used like an RTT signal from a newly acknowledged
+packet.  It MAY change the connection’s default RTT if no RTT measurements
+have been taken.
+
+### Coalesced Packets
+
+Despite loss recovery being separate for different packet number spaces,
+the ability to send a coalesced packet enables faster recovery with small,
+and sometimes no overhead.  The acknowledgement of a coalesced packet
+allows QUIC recovery to use early retransmit to determine if any prior
+packets in that space were lost without waiting for timeouts.
+
+This optimization is particularly useful when:
+
+ * Retransmitting the client’s Initial, which must be padded to a full
+   sized packet, so the datagram typically has extra space to retransmit
+   some outstanding 0-RTT data.
+ * The server sends 1-RTT data soon after the final handshake flight
+   (containing the server Finished) and can proactively retransmit an empty
+   CRYPTO_HS frame bundled with one or more 1-RTT packets.
```

Comment:
The server is not allowed to use 1-RTT keys before receiving the CFIN.

Reply:
Isn't this about the server sending packets?

A server can 1-RTT keys for *sending* packets before receiving ClientFinished. What is not allowed at that moment is using 1-RTT keys for *receiving* packets.

Reply:
It's not equally dangerous because the client is totally anonymous at the point where you are sending 0.5-RTT data, whereas you are acting on the 1-RTT data from the client.

Reply:
I don't think this is that valuable and it's confusing, so I'm removing this example case.